File: symfony/src/Service/Marketplace/CouponService.php
<?php

namespace App\Service\Marketplace;

use App\Entity\Marketplace\Coupon;
use App\Repository\Marketplace\CouponRepository;

class CouponService
{
    public function __construct(
        private CouponRepository $repository,
    ) {}

    public function create(array $data): Coupon
    {
        throw new \LogicException('Not implemented');
    }

    public function update(Coupon $entity, array $data): Coupon
    {
        throw new \LogicException('Not implemented');
    }

    public function isValid(int $id): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Coupon not found: ' . $id);
        $result = $entity->isValid();
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function isApplicableToOrder(int $id, $orderTotal): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Coupon not found: ' . $id);
        $result = $entity->isApplicableToOrder($orderTotal);
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function redeem(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Coupon not found: ' . $id);
        $entity->redeem();
        $this->repository->save($entity, flush: true);
    }

    public function deactivate(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Coupon not found: ' . $id);
        $entity->deactivate();
        $this->repository->save($entity, flush: true);
    }
}

File: rails/app/services/marketplace/coupon_service.rb
module Marketplace
  class CouponService
    def initialize(repository = Coupon)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def is_valid(id)
      instance = Coupon.find(id)
      result = instance.is_valid()
      instance.save!
      result
    end

    def is_applicable_to_order(id)
      instance = Coupon.find(id)
      result = instance.is_applicable_to_order(order_total)
      instance.save!
      result
    end

    def redeem(id)
      instance = Coupon.find(id)
      instance.redeem()
      instance.save!
    end

    def deactivate(id)
      instance = Coupon.find(id)
      instance.deactivate()
      instance.save!
    end
  end
end
